File: HactoberFest/Python/N_Queens.py
def is_safe(board, row, col, n):
    for i in range(col):
        if board[row][i] == 1:
            return False
    
    for i, j in zip(range(row, -1, -1), range(col, -1, -1)):
        if board[i][j] == 1:
            return False
    
    for i, j in zip(range(row, n, 1), range(col, -1, -1)):
        if board[i][j] == 1:
            return False
    
    return True

def solve_nqueens_util(board, col, n):
    if col >= n:
        return True
    
    for i in range(n):
        if is_safe(board, i, col, n):
            board[i][col] = 1
            
            if solve_nqueens_util(board, col + 1, n):
                return True
            
            board[i][col] = 0
    
    return False

def solve_nqueens(n):
    board = [[0 for _ in range(n)] for _ in range(n)]
    
    if not solve_nqueens_util(board, 0, n):
        print("Solution does not exist")
        return
    
    for i in range(n):
        for j in range(n):
            print(board[i][j], end=" ")
        print()

n = int(input())
solve_nqueens(n)

File: Hacktoberfest-2023/Programming/C++/Backtracking/Nqueens.cpp
#include<iostream>
#include<vector>
#include<algorithm>
using namespace std;
//This is a solution for leetcode based c++ problem on the famed Nqueens backtracking problem

class Solution {
public:
    bool isSafe(vector<string>&r,int row,int col,int n)
    {
        for(int i=0;i<col;i++)
        {
            if(r[row][i]=='Q')
               return false;
        }
        for(int i=row,j=col;i>=0 && j>=0; i--,j--)
        {
            if(r[i][j]=='Q')
               return false;
        }
        for(int i=row,j=col;i<n && j>=0;i++,j--)
        {
            if(r[i][j]=='Q')
               return false;
        }
        return true;
    }
    void boardSolveUtil(vector<vector<string>>&board,int col,int n,vector<string>&r)
    {
        int i;
        if(col>=n)
        {
            board.push_back(r);
            return;
        }
        for(i=0;i<n;i++)
        {
            if(isSafe(r,i,col,n))
            {
                r[i][col]='Q';
                boardSolveUtil(board,col+1,n,r);
                r[i][col]='.';
            }
        }
    }
    vector<vector<string>> solveNQueens(int n) 
    {
        vector<vector<string>>board;
        string temp="";
        for(int i=0;i<n;i++)
           temp+='.';
        vector<string>r(n,temp);
        boardSolveUtil(board,0,n,r);
        return board;
        
        
        
    }
};
int main()
{
    Solution obj; 
    vector<vector<string>>board=obj.solveNQueens(4);
    for(int i=0;i<4;i++)
    {
        for(int j=0;j<4;j++)
          cout<<board[i][j]<<" ";
        cout<<endl;
    }
}


//i/o format:
/*
Input: n = 4
Output: [[".Q..","...Q","Q...","..Q."],["..Q.","Q...","...Q",".Q.."]]
*/